Frontend_Web_Development_React_Angular_Vue/React/Week 2/Components.md

```diff
@@ -0,0 +1,110 @@
+```
+In this example:
+- we have created two components, `Greet` and `App`. The first component is `Greet` which is taking a `props` object as an argument and then returning some JSX.
+- Notice the strange {props.name} syntax: this is how you output JavaScript expressions in JSX. An expression is anything that can be the right side of an assignment operator in JavaScript, e.g. const x = <anything that can go here>. If you take away the {} it will literally output props.name to the DOM.
+- And then we have a `App`, which is a Parent component. It's a component that has other components as its children.
+- `Greet` is more flexible component because it accepts some props from its parent component.
+- **Props are variables that a parent (App) passes to its children (the instances of Greet.)**. Now each one can be different! Now that is far more useful than the previous one, since this Greet component can now used to greet any specific person or group of people. Previously It was just `Hello World!`.
+- In `App`, inside the body of return statement, we have a div. This is a container. It's a container for our Greet components, we can now use our Greet components as children of our App component and it accepts props called `name` and it is written like `<Greet name="John">`
+
+> Now the above code is not going run for now because components needs to be export and imported. which is explained in the next lesson 2.4 `import, export, modules`. After that you will able to run this code.
+
+And you will output like this:
+```yml
+Hello John! Welcome to GWOC!
+Hello Jimmy! Welcome to GWOC!
+Hello Jack! Welcome to GWOC!
+```
+This is a complete Information about React Components. Now you should practise by your hand to create some components and pass some props to it and try to render it, you will be able to understand the concept of components. Okay let's disscuss the types of it.
+
+## Types of Components
+
+There are two types of components in React.
+1. `function` components - these are the ones that are written in javascript functions. A function component must return markup.
+example:
+```js
+const App = () => {
+    return (
+        <h1>Hello World!</h1>
+    );
+}
+```
+In the above example we use a Javascript Arrow function to return the markup, we can also write it like this:
+```js
+function App() {
+    return (
+        <h1>Hello World!</h1>
+    );
+}
+```
+The above one is also a function component, which is written in pure Javascipt function.
+after that we can render it to the DOM using `ReactDOM.render`. This is the best practise for writing React Components, most of the React developer follow this practise. 
+
+The output will be:
+
+```html
+Hello World!
+```
+
+2. `class` components - class components are written in ES6 classes. Every class component extends `React.Component`. Every class component must have a render method that returns some sort of JSX / markup / call to `React.createElement`.
```

Comment:
The statement "Every class component extends React.Component" is incorrect. We can make our own class and inherit from that class, to be able to use the methods defined in that parent class.